File: 프로그래머스/2/131127. 할인 행사/할인 행사.cpp
#include <string>
#include <vector>
#include <map>
#include <algorithm>
#include <queue>
using namespace std;

int solution(vector<string> want, vector<int> number, vector<string> discount) {
    int answer = 0;
    	queue<string> que;
	map<string, int> all;
	int cnt = 0;
	
	for (auto n : number)
	{
		cnt += n;
	}
	for (int i = 0; i < want.size(); i++)
	{
		all.insert({ want[i],number[i] });
	}

	for (auto d : discount)
	{
		
		if (find(want.begin(), want.end(), d) == want.end())
		{
			que = {};
            for (int i = 0; i < want.size(); i++)
            {
                all[want[i]] = number[i];
            }
			continue;
		}
		

		if (que.size() == cnt)
		{
			string p = que.front();
			que.pop();
			all[p]++;
		}
		que.push(d);
		all[d]--;

		if (que.size() == cnt)
		{
			int chk = 0;
			for (auto m : all)
			{
				if (m.second == 0) chk++;
			}

			if(chk == want.size())
			{
				answer++;
			}
		}
	}
    return answer;
}

File: 프로그래머스/2/131127. 할인 행사/할인 행사.cs
using System;
using System.Linq;
using System.Collections.Generic;
public class Solution {
    public int solution(string[] want, int[] number, string[] discount) {
int answer = 0;
Dictionary<string, int> dict = new Dictionary<string, int>();
Queue<string> queue = new Queue<string>();

for (int i = 0; i < want.Length; i++)
{
    dict[want[i]] = number[i];
    
}
Dictionary<string, int> dict2 = new Dictionary<string, int>(dict);


for (int i = 0; i < discount.Length; i++)
{
    if (!dict.ContainsKey(discount[i]))
    {
        dict2 = new Dictionary<string, int>(dict);
        queue.Clear();
        continue;
    }


    if (queue.Count < 10)
    {
        queue.Enqueue(discount[i]);
        dict2[discount[i]] -= 1;
    }
    else if (queue.Count == 10)
    {
        queue.Enqueue(discount[i]);
        dict2[queue.Dequeue()] += 1;
        dict2[discount[i]] -= 1;
    }


    bool chk = true;
    foreach (var item in dict2)
    {
        if (item.Value != 0)
        {
            chk = false;
            break;
        }
    }
    if (chk) answer++;

}
        
        return answer;
    }
}